eegnb/experiments/Experiment.py

```diff
@@ -0,0 +1,133 @@
+from eegnb.devices.eeg import EEG
+
+class Experiment:
+
+    def __init__(self, exp_name, duration, eeg, save_fn, n_trials, iti, soa, jitter):
+        """ Anything that must be passed as a minimum for the experiment should be initialized here """
+
+        self.exp_name = exp_name
+        self.instruction_text = """\nWelcome to the {} experiment!\nStay still, focus on the centre of the screen, and try not to blink. \nThis block will run for %s seconds.\n
+        Press spacebar to continue. \n""".format(self.exp_name)
+        self.duration = duration
+        self.eeg = eeg
+        self.save_fn = save_fn
+        self.n_trials = n_trials
+        self.iti = iti
+        self.soa = soa
+        self.jitter = jitter
+        
+    @abstractmethod
+    def load_stimulus(self):
+        """ Needs to be overwritten by specific experiment """
+        raise NotImplementedError
+
+    @abstractmethod
+    def present_stimulus(self):
+        raise NotImplementedError
+
+    def setup(self):
+
+        self.record_duration = np.float32(self.duration)
+        self.markernames = [1, 2]
+        
+        # Setup Trial list -> Common in most (csv in Unicorn)
+        self.parameter = np.random.binomial(1, 0.5, self.n_trials)
+        self.trials = DataFrame(dict(parameter=self.parameter, timestamp=np.zeros(self.n_trials)))
+
+        # Setup Graphics 
+        self.mywin = visual.Window([1600, 900], monitor="testMonitor", units="deg", fullscr=True) 
+        
+        # Needs to be overwritten by specific experiment
+        self.stim = self.load_stimulus()
+        
+        # Show Instruction Screen
+        self.show_instructions()
+
+        if self.eeg:
+            if save_fn is None:  # If no save_fn passed, generate a new unnamed save file
+                random_id = random.randint(1000,10000)
+                self.save_fn = generate_save_fn(self.eeg.device_name, "visual_n170", random_id, random_id, "unnamed")
+                print(
+                    f"No path for a save file was passed to the experiment. Saving data to {self.save_fn}"
+                )
+    
+    def present(self):
+        """ Do the present operation for a bunch of experiments """
+
+        self.setup()
+    
+        # Start EEG Stream, wait for signal to settle, and then pull timestamp for start point
+        if self.eeg:
+            self.eeg.start(self.save_fn, duration=self.record_duration + 5)
+
+        start = time()
+        
+        # Iterate through the events
+        for ii, trial in self.trials.iterrows():
+          
+            # Intertrial interval
+            core.wait(self.iti + np.random.rand() * self.jitter)
+
+            # Some form of presenting the stimulus - sometimes order changed in lower files like ssvep
+            self.present_stimulus(ii)
+
+            # Offset
+            core.wait(self.soa)
+            self.mywin.flip()
+            if len(event.getKeys()) > 0 or (time() - start) > self.record_duration:
```

Comment:
Also as discussed, this part (of the original code) has always looked a bit ugly to me. The alternative is to pre-calculate all of the itis, soas, jitters in advance, and then it will be known before the loop starts what the duration will be, and so the decision of whether or it is the duration param or the completion of the trial set will be clear and not determined on the fly at the end. 

However, I suggest we have a stab at this small adjustment at a later point, it isn't the focus or priority of the current PR.